compiler/rustc_infer/src/infer/error_reporting/note_and_explain.rs

```diff
@@ -77,49 +77,87 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
+                        let parent = p_def_id
                             .as_local()
-                            .map(|id| hir.local_def_id_to_hir_id(id))
-                            .and_then(|id| tcx.hir().find_parent(id))
-                            .as_ref()
-                            .and_then(|node| node.generics())
+                            .and_then(|id| {
+                                let local_id= hir.local_def_id_to_hir_id(id);
+                                let generics = tcx.hir().find_parent(local_id)?.generics()?;
+                                Some((id, generics))
+                            });
+                        if let Some((local_id, generics)) = parent
                         {
                             // Synthesize the associated type restriction `Add<Output = Expected>`.
                             // FIXME: extract this logic for use in other diagnostics.
                             let (trait_ref, assoc_substs) = proj.trait_ref_and_own_substs(tcx);
-                            let path =
-                                tcx.def_path_str_with_substs(trait_ref.def_id, trait_ref.substs);
                             let item_name = tcx.item_name(proj.def_id);
                             let item_args = self.format_generic_args(assoc_substs);
 
-                            let path = if path.ends_with('>') {
-                                format!(
-                                    "{}, {}{} = {}>",
-                                    &path[..path.len() - 1],
-                                    item_name,
-                                    item_args,
-                                    p
-                                )
+                            // Here, we try to see if there's an existing
+                            // trait implementation that matches the one that
+                            // we're suggesting to restrict. If so, find the
+                            // "end", whether it be at the end of the trait
+                            // or the end of the generic arguments.
+                            let mut matching_span = None;
+                            let mut matched_end_of_args = false;
+                            for bound in generics.bounds_for_param(local_id) {
+                                let potential_spans = bound
+                                    .bounds
+                                    .iter()
+                                    .filter_map(|bound| {
+                                        let bound_trait_path = bound.trait_ref()?.path;
+                                        let def_id = bound_trait_path.res.opt_def_id()?;
+                                        let generic_args = bound_trait_path.segments.iter().last().map(|path| path.args());
+                                        (def_id == trait_ref.def_id).then_some((bound_trait_path.span, generic_args))
+                                    })
```

Comment:
What happens if we have like...

```
T: Trait<A> + Trait<B>
```

and then we suggest to restrict `T: Trait<B, Output = C>`

What will that do?

Reply:
Wouldn't the associated type prevent multiple trait impls on the same type anyways? So `Trait<A> + Trait<B>` would error for other reasons before suggesting to constrain the bounds?

Reply:
e.g.

```
trait Foo<T> {
    type Output;

    fn bar() -> Self::Output;
}

fn test<T>() -> i32 where T: Foo<()>, T: Foo<i32> {
    <T as Foo<()>>::bar()
}
```

Reply:
But I guess that already suggests `.. , Output = i32>` correctly